odoo/tools/safe_eval.py

```diff
@@ -238,131 +1348,169 @@ def const_eval(expr):
+    """Allow an instance to be used in the evaluation context
 
-    This can be used to e.g. evaluate
-    an OpenERP domain expression from an untrusted source.
+    Args:
+        t (type): The instance to be allowed
 
-    :param filename: optional pseudo-filename for the compiled expression,
-                     displayed for example in traceback frames
-    :type filename: string
-    :throws TypeError: If the expression provided is a code object
-    :throws SyntaxError: If the expression provided is not valid Python
-    :throws NameError: If the expression provided accesses forbidden names
-    :throws ValueError: If the expression provided uses forbidden bytecode
+    Returns:
+        type: The instance that is allowed
     """
-    if type(expr) is CodeType:
+    global _wrapped_instances  # noqa: PLW0603 (Reason: we need to modify the global variable, check docstring) # pylint: disable=global-statement
+
+    _wrapped_instances += (t,)
+    return t
+
+
+def allow_type(t: type) -> type:
+    """Allow a type to be used in the evaluation context
+
+    Args:
+        t (type): The type to be allowed
+
+    Returns:
+        type: The type that is allowed
+    """
+    global _wrapped_types  # noqa: PLW0603 (Reason: we need to modify the global variable, check docstring) # pylint: disable=global-statement
+
+    _wrapped_types = _wrapped_types + (t,)
+    return allow_instance(t)
+
+
+def test_python_expr(expr: str, mode: str = "eval") -> str | bool:
+    """Test a Python expression before evaluating it
+
+    Args:
+        expr (str): A Python expression
+        mode (str): The mode of the expression (eval, exec)
+
+    Returns:
+        str | bool: An error message if the expression is not safe, False otherwise
+    """
+    try:
+        wrapped_expr = ast.unparse(CodeChecker().visit(ast.parse(expr)))
+        compile(wrapped_expr, "<sandbox>", mode)
+    except (TypeError, NameError, SyntaxError, ValueError) as err:
+        if len(err.args) >= 2 and len(err.args[1]) >= 4:
+            error = {
+                "message": err.args[0],
+                "filename": err.args[1][0],
+                "lineno": err.args[1][1],
+                "offset": err.args[1][2],
+                "error_line": err.args[1][3],
+            }
+
+            msg = f"{type(err).__name__} : {error['message']} at line {error['lineno']}\n{error['error_line']}"
+        else:
+            msg = f"{err}"
+        return msg
+    return False
+
+
+def safe_eval(
+    expr: str,
+    globals_dict: Optional[dict[str, Any]] = None,
+    locals_dict: Optional[dict[str, Any]] = None,
+    mode: str = "eval",
+    filename: Optional[str] = None,
+    locals_builtins: bool = False,
+    sandbox_instances: OptionalTuple[type] = (),
+    sandbox_types: OptionalTuple[type] = (),
+    ast_subset: (
+        frozenset[Type[ast.AST]] | tuple[Type[ast.AST]]
+    ) = _ALLOWED_NODES,  # pylint: disable=bad-whitespace
+) -> Any:
+    def safe_unwrapper(**kwargs) -> Any:
```

Comment:
as discussed on discord, I'm really not a fan of that for a lot of reasons... (performance, security, recursion bug that can happen...) we can try to unwrap only literal types (with literal eval and repr) (and use id instead of record, or strings instead of date, like we already do)